### PR TITLE
Adds a warm-start mechanism to PyOperon

### PR DESCRIPTION
**Pull request by bmcnns**
This pull request adds support for warm start training in the PyOperon scikit-learn bindings, allowing users to call .fit() incrementally. This brings PyOperon in line with other modern frameworks like PySR and Scikit-learn, which already support incremental updates.

By enabling warm starts, users can continue training from a previously fitted model. It also allows users to circumvent the constraints of the callback method  #18.

This feature is particularly important for research scenarios where training without batching can take several days. With warm start support and horizontal batching, experiments that previously took up to 20 days can now be completed far more efficiently. A recent paper, ["A Survey on Batch Training in Genetic Programming", by Liah Rosenfeld and Leonardo Vanneschi](https://link.springer.com/article/10.1007/s10710-024-09501-6), highlights the potential of batching strategies for improving the scalability of genetic programming.

The underlying support for warm starts was added in a corresponding pull request to the Operon core library, and this PR ensures that the Python bindings can take full advantage of that functionality.

This also addresses a feature request in PyOperon #22.

### CODE REVIEW

**Comment by bmcnns** [comment]
I've added an additional method to the C++ implementation called RestoreIndividuals in order to provide a minimal workaround for a new GeneticProgrammingAlgorithm/NSGA2 algorithm object being created for every fit call. Rather than keeping the same GP object after each call, I wanted it to be possible to change some class data such as the Problem for the use-case of batching between calls.

File: pyoperon/sklearn.py [hunk 1]
```diff
@@ -545,10 +548,14 @@ def fit(self, X, y):
         sorter = None if single_objective else op.RankSorter()
         gp     = op.GeneticProgrammingAlgorithm(config, problem, tree_initializer, coeff_initializer, generator, reinserter) if single_objective else \
                  op.NSGA2Algorithm(config, problem, tree_initializer, coeff_initializer, generator, reinserter, sorter)
-        rng    = op.RandomGenerator(np.uint64(config.Seed))
 
-        gp.Run(rng, None, self.n_threads)
+        if self.warm_start and getattr(self, "is_fitted_", False):
```
Comment (foolnotion) [inline pyoperon/sklearn.py:552]: Are we sure this is the correct way to check? The sklearn docs recommend using [`check_is_fitted`](https://scikit-learn.org/stable/modules/generated/sklearn.utils.validation.check_is_fitted.html). Also, shouldn't we check if `is_fitted_` is `True` ? `fit` must have been previously called at least once in order for `self.individuals_` to have a value, right?

Comment (bmcnns) [reply pyoperon/sklearn.py]: I've made some changes to make it more clear but I think this is the right approach.
If we use 'check_is_fitted' to throw an exception (as done in the 'predict' method) then we would need to add an additional setter method for 'warm_start' to enable it after the model 'is_fitted'. This would require an additional manual function call.

Instead, I think it's better to allow the 'warm_start' flag to be enabled even if the estimator hasn't been fit before and to treat it like a cold start for the initial call.
